### ISSUE TITLE
Conan package manager support (using cross building)

### ISSUE THREAD

**Issue by RonAmihai** [zephyrproject-rtos/zephyr#31745]
**Is your feature request related to a problem? Please describe.**
[Conan package manager](https://conan.io/) is a clever way of managing C/C++ project's 3rd party packages/dependencies. 
It can also be used for embedded development using cross building, which requires to manage Conan profiles according to the desired host/target arch platforms and their corresponding toolchains. 

There are many Conan packages which can be compiled successfully with Zephyr and make the development far better.

Configuring Conan manually, especially in case of embedded development & cross building, can be cumbersome. 
Using Zephyr-SDK alongside Conan can be even harder - because of the need to manually configure many profiles for each toolchain & run Conan with the right profile before building for the desired arch.

**Describe the solution you'd like**
Adding Conan support to Zephyr, including appropriate config option, CMake support and internal-predefined conan profiles for every supported Zephyr toolchain. 

The goal is to be able to integrate Zephyr application with Conan, using nothing but a `conanfile.txt` (for listing packages/dependencies for the application) and enable some Conan configs (like `CONFIG_USE_CONAN`). 
It shouldn't even require the user to configure Conan in the project's `CMakeLists.txt`, just add `conanfile.txt` and enable Conan through a Zephyr config.

Zephyr's tree will take care of running Conan onto the project's build directory using the right Conan profile (according to the current Zephyr toolchain, whether it is SDK or specific) & configure/build the Zephyr application afterwards.

**Describe alternatives you've considered**
Configure Conan manually (including modification of the Zephyr app project's `CMakeLists.txt`, adding multiple profiles to support each toolchain/arch, and running Conan from scratch before building the project for different boards)

**Comment by carlescufi** [comment]
@RonAmihai Thanks for the suggestion. Since, AFAIK, we don't have active Conan users so far, could you perhaps send a Draft PR with the files and changes required to achieve that support? Would the SDK also require changes?

**Comment by RonAmihai** [comment]
> @RonAmihai Thanks for the suggestion. Since, AFAIK, we don't have active Conan users so far, could you perhaps send a Draft PR with the files and changes required to achieve that support? Would the SDK also require changes?

The SDK won't require any change in my opinion. 

The only required thing to modify is actually Zephyr's CMake tree, with some side configuration files (mostly Conan profiles for each supported toolchain/arch & maybe some Conan scripts). 

Maybe the Conan install command can be separated into some `west` logic, but this can complicate regular/clean CMake usage with Zephyr, so I think It's better to "force" `conan install --build=missing` as part of the CMake tree by default if Conan support is enabled for the project.

I would like to try helping to Darft PR Conan support soon

**Comment by tim-hilt** [comment]
I‘m also interested in this. However, wouldn’t this PR be better suited for the conan-center repository? It shouldn’t be Zephyrs concern, but rather Conan’s!

**Comment by Ju1He1** [comment]
usually the conan center only hosts 3rd party recipes (i.e. a recipe to build spdlog on any platform).
The conan integration must be located in the consumer project (in this case zephyr).

Our non zephyr projects use [cmake-conan](https://github.com/conan-io/cmake-conan) to have an easy to use conan integration.
Maybe you guys can integrate conan using cmake-conan  :)

**Comment by jasal82** [comment]
+1 interest

**Comment by jankap** [comment]
There's an interesting article about Conan and Zephyr: https://ignitarium.com/streamlining-dependency-management-in-embedded-systems-with-conan-and-zephyr-rtos/

So it seems others are thinking about that, too.

Disclaimer: I'm just searching for Conan + Zephyr solutions, I'm not affiliated with ignitarium

**Comment by jasal82** [comment]
@jankap Thanks for the link. We did something very similar, however, we wanted to re-use existing Conan packages without having to modify their conanfile and/or CMakeLists.txt. So our approach is to use a `zephyr-sdk` package (similar to the one mentioned in the article) as toolchain package that is injected via the Conan host profile and generates a user toolchain file. That user toolchain file is populated with the compiler and linker flags and the generated headers required for making CMake use the Zephyr toolchain instead of a normal compiler. This enables us to make Conan build all transitive dependencies with the Zephyr toolchain and then integrate all those dependencies in a Zephyr application project via `find_package()` and `target_link_libraries()`. So this can basically use any existing Conan package as requirement, provided that it is able to create a static library (because this is the only artifact that can be injected into the Zephyr build, apart from the source files themselves). I also managed to remove the requirement for creating a Python venv by making Conan bootstrap one itself.

I must add, however, that we're still not satisfied with this solution as it requires a Zephyr dummy build as part of the `zephyr-sdk` package installation so that the compiler flags and generated headers can be retrieved. And this dummy build also needs the correct Zephyr application configuration (`prj.conf`, device tree overlay, etc) because they can affect the compiler flags and the generated headers. This configuration must be injected into the Conan build using `conf` settings that represent the path to the configuration files (and thus require custom package ID calculation). Last but not least Zephyr needs help with finding the correct order for linking libraries in the final application build because some Conan components (those with actual dependencies to Zephyr headers) depend on the Zephyr `kernel` library, but static libraries added with `target_link_libraries()` are always added after the `kernel` library. However, they need to go before it, otherwise the unused symbols in the `kernel` lib have already been removed by the linker. I found no way to define that dependency in the Zephyr app CMakeLists. It must be defined in the Conan-generated CMake module, so the respective conanfile needs a `self.cpp_info.system_libs.append("kernel")`.

To sum up, it would really be great if Zephyr would respect CMake toolchain files. Together with some minor changes in Conan this would make the integration much easier.
